samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -0,0 +1,221 @@
+    // model generation - NEEDS TO BE REVIEWED
+    JavaSystemConfig systemConfig = new JavaSystemConfig(this.config);
+    Map<String, SystemAdmin> systemAdmins = new HashMap<>();
+    for (String systemName: systemConfig.getSystemNames()) {
+      String systemFactoryClassName = systemConfig.getSystemFactory(systemName);
+      if (systemFactoryClassName == null) {
+        String msg = String.format("A stream uses system %s, which is missing from the configuration.", systemName);
+        log.error(String.format(msg));
+        throw new SamzaException(msg);
+      }
+      SystemFactory systemFactory = Util.getObj(systemFactoryClassName);
+      systemAdmins.put(systemName, systemFactory.getAdmin(systemName, this.config));
+    }
+
+    return new StreamMetadataCache(Util.<String, SystemAdmin>javaMapAsScalaMap(systemAdmins), 5000, SystemClock.instance());
+  }
+
+  @Override
+  public void start() {
+    zkController.register();
+  }
+
+  public void cleanupZk() {
+    zkUtils.deleteRoot();
+  }
+
+  @Override
+  public void stop() {
+    zkController.stop();
+  }
+
+  @Override
+  public boolean awaitStart(long timeoutMs)
+      throws InterruptedException {
+    return containerController.awaitStart(timeoutMs);
+  }
+
+  @Override
+  public int getProcessorId() {
+    return processorId;
+  }
+
+  @Override
+  public JobModel getJobModel() {
+    return newJobModel;
+  }
+
+  //////////////////////////////////////////////// LEADER stuff ///////////////////////////
+  @Override
+  public void onBecomeLeader() {
+    log.info("ZkJobCoordinator::onBecomeLeader - I became the leader!");
+
+    List<String> emptyList = new ArrayList<>();
+
+    // actual actions to do are the same as onProcessorChange()
+    debounceTimer.scheduleAfterDebounceTime(ScheduleAfterDebounceTime.ON_PROCESSOR_CHANGE,
+        ScheduleAfterDebounceTime.DEBOUNCE_TIME_MS, () -> onProcessorChange(emptyList));
+  }
+
+  @Override
+  public void onProcessorChange(List<String> processorIds) {
+    log.info("ZkJobCoordinator::onProcessorChange - Processors changed! List: " + Arrays.toString(processorIds.toArray()));
+    generateNewJobModel();
+  }
+
+  @Override
+  public void onNewJobModelAvailable(final String version) {
+    newJobModelVersion = version;
+    log.info("pid=" + processorId + "new JobModel available");
+    // stop current work
+    containerController.stopContainer();
+    log.info("pid=" + processorId + "new JobModel available.Container stopped.");
+    // get the new job model
+    newJobModel = zkUtils.getJobModel(version);
+    log.info("pid=" + processorId + "new JobModel available. ver=" + version + "; jm = " + newJobModel);
+
+    String currentPath = zkUtils.getEphemeralPath();
+    String zkProcessorId = keyBuilder.parseIdFromPath(currentPath);
+
+    // update ZK and wait for all the processors to get this new version
+    barrier.waitForBarrier(version, String.valueOf(zkProcessorId), new Runnable() {
```

Comment:
Do we need some timeout mechanism here? I commented in the previous barrier PR too. If a processor failed during previous stopContainer, then the Barrier will not be reached in the future. I think this can happen in real jobs.